libsolidity/interface/CompilerStack.cpp

```diff
@@ -1672,6 +1674,15 @@ std::string CompilerStack::createMetadata(Contract const& _contract, bool _forIR
 			details["yulDetails"]["stackAllocation"] = m_optimiserSettings.optimizeStackAllocation;
 			details["yulDetails"]["optimizerSteps"] = m_optimiserSettings.yulOptimiserSteps + ":" + m_optimiserSettings.yulOptimiserCleanupSteps;
 		}
+		else if
+		(
+			OptimiserSuite::isEmptyOptimizerSequence(m_optimiserSettings.yulOptimiserSteps) &&
+			OptimiserSuite::isEmptyOptimizerSequence(m_optimiserSettings.yulOptimiserCleanupSteps)
+		)
+		{
+			details["yulDetails"] = Json::objectValue;
+			details["yulDetails"]["optimizerSteps"] = m_optimiserSettings.yulOptimiserSteps + ":" + m_optimiserSettings.yulOptimiserCleanupSteps;
+		}
```

Comment:
> Going with @ekpyron's reasoning that these settings in metadata should reflect what compiler does rather than how input looked like, we could just always include `yulDetails` and all of its fields, regardless of the value of the `yul` flag. This actually better reflects reality now because these settings do have effect even with `yul: false`, we just don't allow certain combinations. Then we would not have to make any assumptions about defaults and assertions could be safely omitted.
> 
> One small issue I see though is that we'd be outputting `yulDetails` in cases where we'd not accept it as input so constructing the input from metadata would become a tiny bit more cumbersome.
> 
> It would also change how settings for `yul: false` look like in metadata (they did not have `yulDetails`, now they always would). The new form would be equivalent so I don't think it counts as breaking, but we should probably make sure that Sourcify can handle it if we decide to do that (pinging @kuzdogan).

Perfectly fine with me, but it would be nice to have input from both @ekpyron and @kuzdogan, given that we're targeting for  a release tomorrow.

Reply:
I don't see much reason for making the metadata settings unnecessarily verbose. If you can achieve the effect *without* specifying details, since it's implied by default by the previous settings, then we can and should use the minimal settings.
That's also what the comparison with ``OptimizerSettings::minimal()`` and ``OptimizerSettings::standard()`` achieves above.

That being said, what's *really* important is the following two properties:
- The settings in the metadata can be fed back to the compiler as they are without any change to result in the same compiler behavior (that's the only reason we can change how any of this works in the first place without it being breaking)
- Whenever possible metadata should contain the *canonical* version of the settings, so if there is multiple combinations of settings that are guaranteed to achieve the exact same compiler behaviour we should choose *one* of them to enter to the metadata in *all* such cases. Ideally that'd be the simplest one (meaning not specifying details if the behavior can be achieved without specifying them)

Reply:
> The settings in the metadata can be fed back to the compiler as they are without any change to result in the same compiler behavior

Can they always be fed back though? I was under impression that this is not always possible, and that settings may require translation/adjustment in some cases anyway. 

But, well, maybe I just confused this with the fact that metadata as a whole is not usable as input as is... If settings can be used unmodified then I guess it makes sense to keep that property.

Still, we should really have these principles documented somewhere (comments? docs?). This can be done in different ways and the correct one is not always obvious to me.

In any case, we just need the asserts here.